[SRVKS-957] Add section on resolving Serverless operator upgrade failure (#58238)
openshift/openshift-docs: modules/serverless-resolving-operator-upgrade-failure.adoc

// Module included in the following assemblies:
//
// * /serverless/admin_guide/serverless-upgrades.adoc

:_content-type: PROCEDURE
[id="serverless-resolving-operator-upgrade-failure_{context}"]
= Resolving {ServerlessOperatorName} upgrade failure

In some cases, such as when you do manual uninstalls and reinstalls, you might encounter an error when upgrading {ServerlessOperatorName}. In this situation, you need to manually reinstall {ServerlessOperatorName}.

.Procedure

> . Identify the version of {ServerlessOperatorName} that was installed originally using the `project.yaml` files for the different releases.

Reviewer: I'm not very happy to point customers to the "development" project.yaml from our repository. We could probably achieve the same by pointing them to release notes.

If the installed KnativeServing version is v1.5.0 where the version has this pattern: vMAJOR.MINOR.PATCH. Then searching through release notes for MAJOR.MINOR version of Knative Serving will give them the version of Serverless. 
For example, [release notes for 1.26](https://58238--docspreview.netlify.app/openshift-enterprise/latest/serverless/serverless-release-notes.html#serverless-rn-1-26_serverless-release-notes) have this: `OpenShift Serverless now uses Knative Serving 1.5.`

+
For example, the error message during attempted upgrade contains this string:
+
[source,yaml]
----

> The installed KnativeServing version is v0.25.1.

Reviewer: Let's update the versions to something more recent, for example `v1.5.0`.  (which corresponds to Serverless 1.26.0). I will provide examples below for this version.

----
+

> In this case, the KnativeServing version is 0.25.1, which corresponds to {ServerlessOperatorName} 1.19.

Reviewer: ```suggestion
In this case, the KnativeServing version is 1.5.0, which corresponds to {ServerlessOperatorName} 1.26.
```

> . Uninstall {ServerlessOperatorName} and all its install plans.

Reviewer: For this step, I'd point to [this section](https://docs.openshift.com/container-platform/4.12/serverless/removing/removing-serverless-operator.html#olm-deleting-operators-from-a-cluster-using-web-console_removing-serverless-operator) which does exactly what they need: remove subscription, installplans and serverless-operator ClusterServiceVersion.

> . Manually install {ServerlessOperatorName} of version discovered in the first step.

Reviewer: In this step we need to provide an example Subscription otherwise it will be hard to figure this out (one has to specify startingCSV). The example below should be fine:

Create serverless-subscription.yaml with the following content:
```
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: serverless-operator
  namespace: openshift-serverless
spec:
  channel: stable
  name: serverless-operator
  source: redhat-operators
  sourceNamespace: openshift-marketplace
  installPlanApproval: Manual
  startingCSV: serverless-operator.v1.26.0
```
Install the subscription:
```
oc apply -f serverless-subscription.yaml
```

// Module included in the following assemblies:
//
// * /serverless/admin_guide/serverless-upgrades.adoc

:_content-type: PROCEDURE
[id="serverless-resolving-operator-upgrade-failure_{context}"]
= Resolving {ServerlessOperatorName} upgrade failure

> In some cases, such as when you do manual uninstalls and reinstalls, you might encounter an error when upgrading {ServerlessOperatorName}. In this situation, you need to manually reinstall {ServerlessOperatorName}.

Reviewer 2: ```suggestion
You might encounter an error when upgrading {ServerlessOperatorName}, for example, when performing manual uninstalls and reinstalls. If you encounter an error, you must manually reinstall {ServerlessOperatorName}.
```

.Procedure

. Identify the version of {ServerlessOperatorName} that was installed originally by searching in the {ServerlessProductName} Release Notes.
+

> For example, the error message during attempted upgrade contains this string:

Reviewer 2: ```suggestion
For example, the error message during attempted upgrade might contain the following string:
```

+
[source,yaml]

> ----

Reviewer 3: @rh-max, should this be YAML source?

The installed KnativeServing version is v1.5.0.
----
+

> In this case, the KnativeServing `MAJOR.MINOR` version is `1.5`, which is mentioned in the release notes for {ServerlessProductName} 1.26: `OpenShift Serverless now uses Knative Serving 1.5`.

Reviewer 2: ```suggestion
In this example, the KnativeServing `MAJOR.MINOR` version is `1.5`, which is covered in the release notes for {ServerlessProductName} 1.26: `OpenShift Serverless now uses Knative Serving 1.5`.
```

Reviewer 2: I think the title of the Release Notes item would be in italics, not monospace?

> . Uninstall {ServerlessOperatorName} and all its install plans.

Reviewer 2: ```suggestion
. Uninstall {ServerlessOperatorName} and all of its install plans.
```

> . Manually install {ServerlessOperatorName} of version discovered in the first step. To do that, first create a `serverless-subscription.yaml` file:

Reviewer 2: ```suggestion
. Manually install the version of {ServerlessOperatorName} that you discovered in the first step. To install, first create a `serverless-subscription.yaml` file as shown in the following example:
```

+
[source,yaml]
----
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: serverless-operator
  namespace: openshift-serverless
spec:
  channel: stable
  name: serverless-operator
  source: redhat-operators
  sourceNamespace: openshift-marketplace
  installPlanApproval: Manual
  startingCSV: serverless-operator.v1.26.0
----

> . Then, install the subscription:

Reviewer 2: ```suggestion
. Then, install the subscription by running the following command:
```

+
[source,terminal]
----
$ oc apply -f serverless-subscription.yaml
----

> . Upgrade by manually approving the upgrade install plans as they appear.

Reviewer 2: ```suggestion
. Upgrade by manually approving the upgrade install plans as they appear in the terminal window.
```

Reviewer 2: Added this but may not be correct - it seemed to need some clarification, but please ignore if not applicable.
